crates/snapshot/src/lib.rs

```diff
@@ -771,6 +846,170 @@ impl SnapshotRepository {
         }
         Ok(())
     }
+
+    /// Compress the snapshot of the replica with the given `tx_offset`.
+    ///
+    /// If the snapshot is already compressed, it will return the compress type.
+    ///
+    /// If the snapshot is not compressed, it will compress it and return the compress type.
+    ///
+    /// If the snapshot is locked, it will return an [`SnapshotError::CompressLocked`].
+    pub fn compress_snapshot(
+        snapshot_dir: &SnapshotDirPath,
+        tx_offset: TxOffset,
+    ) -> Result<CompressType, SnapshotError> {
+        let snapshot_file = snapshot_dir.snapshot_file(tx_offset);
+        let (snapshot, compress_type) = Snapshot::read_from_file(&snapshot_file)?;
+
+        if compress_type != CompressType::None {
+            log::info!(
+                "Snapshot {snapshot_dir:?} of replica {} is already compressed: {compress_type:?}",
+                snapshot.replica_id
+            );
+            return Ok(compress_type);
+        }
+
+        let lockfile = Lockfile::lock_path(snapshot_dir);
+        if lockfile.try_exists()? {
+            let err = SnapshotError::CompressLocked { tx_offset, lockfile };
+            log::error!("{err}");
+            return Err(err);
+        }
+
+        let _lock = Lockfile::for_file(snapshot_dir)?;
+
+        log::info!(
+            "Compressing snapshot {snapshot_dir:?} of replica {}",
+            snapshot.replica_id
+        );
+
+        // Replace the original file with the compressed one.
+        fn compress(src: &PathBuf) -> io::Result<()> {
+            let mut read = CompressReader::new(o_rdonly().open(src)?)?;
+            if read.compress_type() != CompressType::None {
+                return Ok(()); // Already compressed
+            }
+            let dst = src.with_extension("_tmp");
+            let mut write = CompressWriter::new(o_excl().open(&dst)?, CompressType::Zstd)?;
+            std::io::copy(&mut read, &mut write)?;
+            std::fs::rename(dst, src)?;
```

Comment:
Note that `rename(2)` does not preserve the inode, i.e. all objects in later snapshots that were hardlinked to the original file will remain uncompressed. This means the replica's snapshot tree occupies more space than before.

It would be good thus to employ the same strategy as when creating snapshots: check if the previous snapshot has a compressed version of this file, and hardlink to that instead of freshly compressing it.